src/template.c

```diff
@@ -1517,20 +1518,95 @@ MATCH TemplateDeclaration::deduceFunctionTemplateMatch(
             }
         }
 
+        Expression *farg;
         if (argi >= nfargs)                // if not enough arguments
         {
-            if (fparam->defaultArg)
+            if (!fparam->defaultArg)
+                goto Lnomatch;
+
+            /* Bugzilla 2803: Before the starting of type deduction from the function
+             * default arguments, lock the already deduced parameters into paramscope.
+             * It's necessary to avoid breaking some existing code. See the regression cases
+             * in `runnable/template9.d`.
```

Comment:
I had just specified the issue 2803 test case. It's only 500 lines so finding it would be easy.
But OK, I added acceptable cases in the comment.